### PR TITLE
@feature/support for c# nullable projects #101

### LINKED ISSUE #101: Support for C# Nullable Projects
Hello,
We would like to use Scaffolding Handlebars in a Nullable C# project; as such there needs to be a way to add "?" after fields that are nullable. 

To not maintain a fork separately Ideally looking to have a merge of :
https://github.com/TrackableEntities/EntityFrameworkCore.Scaffolding.Handlebars/pull/99

### PR DESCRIPTION
**Pull request by JohnGoldInc**
Closes #101

### CODE REVIEW

**Comment by JohnGoldInc** [comment]
@tonysneed Really having a hard time with squash and getting it to work right (sorry)

**Comment by tonysneed** [comment]
Don't worry about squash just yet. I can show you how to do that after I've reviewed and approved your PR.

More importantly, you should try to follow each step in the Contributing Guidelines, especially [Step 15](https://github.com/TrackableEntities/EntityFrameworkCore.Scaffolding.Handlebars/blob/master/.github/CONTRIBUTING.md#create-pull-request). But don't worry about that right now either. Just hang tight while I review the PR.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs [hunk 1]
```diff
@@ -2,7 +2,7 @@
 {{#each property-annotations}}
 {{spaces 8}}{{{property-annotation}}}
 {{/each}}
-{{spaces 8}}public {{property-type}} {{property-name}} { get; set; }
+{{spaces 8}}public {{property-type}} {{property-name}} { get; set; }{{#if nullable-reference-types }} = default!;{{/if}}
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:5]: You also need to check `property-isnullable` here, so that `= default!;` is only added when `property-isnullable` is false. For example:
```csharp
public partial class Customer
{
    public string CustomerId { get; set; } = default!;
    public string? ContactName { get; set; }
```

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs [hunk 2]
```diff
@@ -11,9 +11,9 @@
 {{spaces 8}}{{{nav-property-annotation}}}
 {{/each}}
 {{#if nav-property-collection}}
-{{spaces 8}}public virtual ICollection<{{nav-property-type}}> {{nav-property-name}} { get; set; }
+{{spaces 8}}public virtual ICollection<{{nav-property-type}}> {{nav-property-name}} { get; set; }{{#if nullable-reference-types}} = null!;{{/if}}
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:14]: Instead of `null!` let's be consistent and use `default!` throughout.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs [hunk 3]
```diff
@@ -11,9 +11,9 @@
...
 {{else}}
-{{spaces 8}}public virtual {{nav-property-type}} {{nav-property-name}} { get; set; }
+{{spaces 8}}public virtual {{nav-property-type}} {{nav-property-name}} { get; set; }{{#if nullable-reference-types}} = null!;{{/if}}
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:16]: Use `default!`

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpDbContext/Partials/DbSets.hbs [hunk 4]
```diff
@@ -1,3 +1,3 @@
 ﻿{{#each dbsets}}
-{{spaces 8}}public virtual DbSet<{{set-property-type}}> {{set-property-name}} { get; set; }
+{{spaces 8}}public virtual DbSet<{{set-property-type}}> {{set-property-name}} { get; set; }{{#if nullable-reference-types }} = null!;{{/if}}
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpDbContext/Partials/DbSets.hbs:2]: Use `default!`

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs [hunk 5]
```diff
@@ -94,13 +94,17 @@ public string TransformPropertyName(string propertyName)
             foreach (var line in lines)
             {
                 var propTypeInfo = new EntityPropertyInfo(line["property-type"] as string,
-                    line["property-name"] as string);
+                    line["property-name"] as string,
+                    line["property-isnullable"] as bool?
+                    );
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs:99]: Don't think we need to change the ctor transformer.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs [hunk 6]
```diff
@@ -94,13 +94,17 @@ public string TransformPropertyName(string propertyName)
...
                 var transformedProp = ConstructorTransformer?.Invoke(propTypeInfo) ?? propTypeInfo;
 
                 transformedLines.Add(new Dictionary<string, object>
                 {
                     { "property-name", transformedProp.PropertyName },
                     { "property-type", transformedProp.PropertyType },
+                    { "property-isnullable", transformedProp.PropertyIsNullable },
+                    { "nullable-reference-types", line["nullable-reference-types"]}
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs:107]: Same as above. Don't change ctor transformer.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs [hunk 7]
```diff
@@ -119,14 +123,18 @@ public string TransformPropertyName(string propertyName)
             foreach (var property in properties)
             {
                 var propTypeInfo = new EntityPropertyInfo(property["property-type"] as string, 
-                    property["property-name"] as string);
+                    property["property-name"] as string,
+                    property["property-isnullable"] as bool?
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs:127]: Don't need `property-isnullable` for nav properties.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs [hunk 8]
```diff
@@ -119,14 +123,18 @@ public string TransformPropertyName(string propertyName)
...
+                    );
                 var transformedProp = PropertyTransformer?.Invoke(propTypeInfo) ?? propTypeInfo;
 
                 transformedProperties.Add(new Dictionary<string, object>
                 {
                     { "property-type", transformedProp.PropertyType },
                     { "property-name", transformedProp.PropertyName },
-                    { "property-annotations", property["property-annotations"] }
+                    { "property-annotations", property["property-annotations"] },
+                    { "property-isnullable", property["property-isnullable"] },
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs:136]: `property-isnullable` not needed.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs (hunk 2 above)
Comment (JohnGoldInc) [reply src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:14]: @tonysneed kinda feels wrong to use default on navigation props, what about using null! everywhere instead?
The example they give uses null:  https://docs.microsoft.com/en-us/ef/core/miscellaneous/nullable-reference-types

I haven't tested it but I worry what side effect default might have?  null I know will create runtime errors; however I'm worried default might create insidious logic errors (i.e. default to and empty collection deleting records)

Going to have to test this , not finding articles discussing this...

Comment (JohnGoldInc) [reply src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:14]: @tonysneed Tested it Never mind, your're right

**Comment by tonysneed** [comment]
Yes, `default` is better. It equates to `null` for reference types and zeroed out bits for value types.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs (hunk 2 above)
Comment (tonysneed) [reply src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:14]: Yes, `default` is better. It equates to `null` for reference types and zeroed out bits for value types.

**Comment by JohnGoldInc** [comment]
@tonysneed  Ok, All Fixed up now; think I came to better naming which explains why I am doing some things. (like why I needed to change HbsEntityTypeTransformationService.cs  )

**Comment by tonysneed** [comment]
I hate to say it, but you should keep `nullable-reference-types` instead of `property-initializes-default`. The reason is that this info is not specific to a property, but rather it is based on `_options?.Value?.EnableNullableReferenceTypes`. Can you please change it back?

Also put `EntityPropertyInfo` back the way it was before. `PropertyIsNullable` was correct.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs [hunk 9]
```diff
@@ -2,7 +2,7 @@
 {{#each property-annotations}}
 {{spaces 8}}{{{property-annotation}}}
 {{/each}}
-{{spaces 8}}public {{property-type}} {{property-name}} { get; set; }{{#if nullable-reference-types }} = default!;{{/if}}
+{{spaces 8}}public {{property-type}} {{property-name}} { get; set; }{{#if property-initializes-default }} = default!;{{/if}}
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/CodeTemplates/CSharpEntityType/Partials/Properties.hbs:5]: Only add `= default!;` when `nullable-reference-types` is true and `property-isnullable` is false. See `Customer` class example [here](https://github.com/TrackableEntities/EntityFrameworkCore.Scaffolding.Handlebars/pull/115#discussion_r418385146).
```
{{#if nullable-reference-types && !property-isnullable }} = default!;{{/if}}
```

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs (hunk 7 above)
Comment (tonysneed) [reply src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs]: Sorry, I think this comment was in error because the code is in `TransformProperties`, not `TransformNavigationProperties`.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/EntityPropertyInfo.cs [hunk 10]
```diff
@@ -15,12 +15,12 @@ public class EntityPropertyInfo
         /// </summary>
         /// <param name="propertyType">Property type.</param>
         /// <param name="propertyName">Property name.</param>
-        /// <param name="propertyIsNullable">Property is nullable.</param>
-        public EntityPropertyInfo(string propertyType, string propertyName, bool? propertyIsNullable)
+        /// <param name="propertyInitializesDefault">Property initializes default "... { get; set; } = default!;"</param>
+        public EntityPropertyInfo(string propertyType, string propertyName, bool? propertyInitializesDefault = null)
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/EntityPropertyInfo.cs:19]: Please revert all changes to `EntityPropertyInfo`. `PropertyIsNullable` is the right property.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/EntityPropertyInfo.cs [hunk 11]
```diff
@@ -36,6 +36,6 @@ public EntityPropertyInfo(string propertyType, string propertyName, bool? proper
         /// <summary>
         /// Property is nullable.
         /// </summary>
-        public bool? PropertyIsNullable { get; set; }
+        public bool? PropertyInitializesDefault { get; set; }
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/EntityPropertyInfo.cs:39]: Make this `bool` not `bool?`. The property is either nullable or not nullable.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsCSharpEntityTypeGenerator.cs [hunk 12]
```diff
@@ -222,8 +220,7 @@ protected override void GenerateProperties(IEntityType entityType)
                     { "property-type", propertyType },
                     { "property-name", property.Name },
                     { "property-annotations",  PropertyAnnotationsData },
-                    { "property-isnullable",  property.IsNullable },
-                    { "nullable-reference-types",  _options?.Value?.EnableNullableReferenceTypes == true }
+                    { "property-initializes-default", _options?.Value?.EnableNullableReferenceTypes == true && ! property.IsNullable }
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsCSharpEntityTypeGenerator.cs:223]: Please revert this change. `property-isnullable` and `nullable-reference-types` are both needed in the `GenerateProperties` method.

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs [hunk 13]
```diff
@@ -124,17 +122,15 @@ public string TransformPropertyName(string propertyName)
             {
                 var propTypeInfo = new EntityPropertyInfo(property["property-type"] as string, 
                     property["property-name"] as string,
-                    property["property-isnullable"] as bool?
-                    );
+                    property["property-initializes-default"] as bool?);
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs:125]: Change this back to `property-isnullable`

File: src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs [hunk 14]
```diff
@@ -124,17 +122,15 @@ public string TransformPropertyName(string propertyName)
...
                 var transformedProp = PropertyTransformer?.Invoke(propTypeInfo) ?? propTypeInfo;
 
                 transformedProperties.Add(new Dictionary<string, object>
                 {
                     { "property-type", transformedProp.PropertyType },
                     { "property-name", transformedProp.PropertyName },
                     { "property-annotations", property["property-annotations"] },
-                    { "property-isnullable", property["property-isnullable"] },
-                    { "nullable-reference-types", property["nullable-reference-types"]}
+                    { "property-initializes-default", transformedProp.PropertyInitializesDefault }
```
Comment (tonysneed) [inline src/EntityFrameworkCore.Scaffolding.Handlebars/HbsEntityTypeTransformationService.cs:133]: Please revert this change.

**Comment by JohnGoldInc** [comment]
@tonysneed Fixed, had to do handlebars syntax as:

`{{#if nullable-reference-types }}{{#unless property-isnullable}} = default!;{{/unless}}{{/if}}`

Because no logical "and" or "not" in handlebars plain

**Comment by JohnGoldInc** [comment]
@tonysneed I checked in my changes to the expected context strings this time with a replacing slashes with double slashes logic.

**Comment by tonysneed** [comment]
@JohnGoldInc OK I think we are ready for the merge. If you allow me to make changes to your branch, I will be glad to rebase and squash your commits.

In the lower right-hand corner of this PR you should see a checkbox labeled **Allow edits and access to secrets by maintainers**. Please check it. Then I will be able to squash and merge your PR. Thanks!

**Comment by JohnGoldInc** [comment]
I just check it but I think since you have merged, it doesn't matter anymore and doesn't stay set :)

Thanks for helping me through, git with forked projects is so much more than what I am used too.

**Comment by tonysneed** [comment]
@JohnGoldInc Thanks very much for your contribution! I will mention you in the next release.

To learn more about the Git workflow I use, you can watch this [online tutorial](https://youtu.be/gxhbH9fzTDo) I created.
